keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+
+  **Mitigation:** Properly document the suggested migration paths with
+  limitations.
+
+## Design Details
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy (along with EndpointSlice controller, within kube-controller-manager
+as the control plane) will support the three standard routing preferences
+(`Default`, `Spread`, `Zone`).
+
+#### `Default` and `Spread`
+* Initially, kube-proxy will treat the `Default` preference the same as
+  `Spread`
+* This leverages existing implementation, requiring no major changes.
+
+#### `Zone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  `Zone` heuristic – the hints will match the zone of the endpoint itself.
+* While it may seem redundant to populate the hints here since kube-proxy can
+  already derive the zone hint from the endpoints zone (as they would be the
+  same), we will still use this for implementation simply because of the reason
+  that it’s easier to implement and provides a better design. Consider an
+  alternative implementation where kube-proxy reads
+  `routingPreference=Zone` field and then constructs the route rules
+  accordingly. This means some extra logic needs to be baked into the kube-proxy
+  which could have just as easily been implemented by an already existing
+  extensibility mechanism (i.e. EndpointSlice hints)
+
+Although this is not an explicit design goal, an implication of the above
+implementation choice means that:
+* The control plane (kube-controller-manager in this case) is only concerned
+  with the field `routingPreference` and populates EndpointSlice hints based
+  on its value 
+* The data plane (kube-proxy in this case) is only concerned with the Hints
+  populated in the EndpointSlice and the fields `internal/externalTrafficPolicy`
+  to make routing decisions.
+* Neither the control plane nor the data plane looks at the others field.
+
+NOTE: The fact that EndpointSlice hints are not expected to be implemented by
+  all data planes is not of concern here, because `routingPreference` is anyways
+  supposed to be implementation dependent. Since we know that the kube-proxy
+  data plane will respect the Hints, it’s sufficient for the kube-proxy and
+  kube-controller-manager based implementations to say that we do support the
+  standard heuristics.
+
+### Changes within kube-proxy
+
+**Present behaviour:** kube-proxy only considers EndpointSlice hints for route
+programming if the `service.kubernetes.io/topology-aware-hints` annotation is
+set to "Auto"
+
+**New behaviour:** Irrespective of what the annotation
+`service.kubernetes.io/topology-aware-hints` or field `routingPreference` are
+set to (or even if they are not set at all), kube-proxy will always consider
+EndpointSlice hints (assuming this feature-gate is enabled).
```

Comment:
It will "consider" hints regardless of `routingPreference` value? That needs clarification. I assume it "considers" them in very different ways for `Spread` vs `Zone`.